demos/meltdown_ud.cc

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#ifndef __aarch64__
+#  error Unsupported architecture. ARM64 required.
+#endif
+
+#include <signal.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Local handler necessary for avoiding local/global linking mismatches on ARM.
+static void local_handler() {
+  asm volatile("b afterspeculation");
+}
+
+static char leak_byte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
```

Comment:
Can we avoid this and use `AddHitAndRecomputeScores` like we did in other Meltdown examples?

Reply:
When I plainly convert it to AddHitAndRecompute, it never works. I can help it with adding a DSB ISB before the undefined instruction and then it works about each second attempt on Cavium. But still it is a radical drop of reliability when the oracle handlers and the data pointer is not accessed beforehand. Also we do not use AddHitAndRecompute in the Meltdown (Meltdown-US-L1) example exactly because of this reason https://github.com/google/safeside/blob/master/demos/meltdown.cc#L75 and there are no other Meltdown examples. Or you have already tried it and it worked somewhere? We use it only in ret2spec.

Reply:
We can probably solve this issue together with the main Meltdown example that you can test also on Intel CPUs.